quinn-proto/src/endpoint.rs

```diff
@@ -702,7 +708,22 @@ impl Endpoint {
     }
 
     /// Reject this incoming connection attempt
+    #[deprecated(since = "0.11.13", note = "use `refuse_reason()` instead")]
     pub fn refuse(&mut self, incoming: Incoming, buf: &mut Vec<u8>) -> Transmit {
+        self.refuse_reason(
+            incoming,
+            Some("connection refused by policy".to_owned()),
+            buf,
+        )
+    }
+
+    /// Reject this incoming connection attempt
+    pub fn refuse_reason(
+        &mut self,
+        incoming: Incoming,
+        reason: Option<String>,
```

Comment:
`Connection::close` takes `Bytes`. Should we do the same here?

Reply:
Why? `ApplicationClose::reason` is (publicly) `Bytes` but `TransportError::reason` is (publicly) `String` -- I feel like `String` is probably more reasonable in terms of the public API, and it doesn't feel like there's a bunch of performance to be had from refcounting error reasons instead of allocating?

Reply:
It gets baked down into `ConnectionClose::reason`, which exactly mirrors `ApplicationClose::reason` as `Bytes`, and is not strictly required to be UTF-8. The involvement of `TransportError` is an artifact of the private `initial_close` API, which we can easily revisit. Most important, though, is just that it would be more consistent with `close`, which is morally the same operation, just at a different time.

Reply:
I agree that there's a case for preferring `String` here, though I don't think it's totally clear cut -- it's nice to leave `SHOULD` stuff to the application when feasible, especially when the costs of doing it wrong are limited.

However, we *already have* a `Bytes` API for this elsewhere. IMO consistency trumps this concern entirely.